feat(grouptombstones): Add 'last_seen' and 'times_seen' columns

priscilawebdev opened the pull request:
**Motivation**
Currently, when an issue is marked to be discarded (via "Delete and discard future events"), we create a GroupTombstone and silently drop any new matching events. While this keeps the issue stream clean, it provides no visibility into how often these discarded issues are still occurring. Without this data, developers can't know if a discarded issue is a frequently recurring problem that might need to be revisited.

**Changes**
These changes improve our tracking of discarded issues ([read more](https://blog.sentry.io/delete-and-discard/)) by adding two new columns to the `GroupTombstone` table:

1.  `times_seen`: This column is incremented each time a new event matches the `GroupTombstone`. It allows us to quantify the frequency of a discarded issue.
2.  `last_seen`: This column stores the timestamp of the most recent event that matched the `GroupTombstone`.

When an event matches a discarded hash, the `EventManager` now updates these fields on the corresponding `GroupTombstone`, providing valuable metadata about ongoing discarded events.

Contributes to https://linear.app/getsentry/issue/TET-595/display-discarded-issue-count-and-timestamp-for-each-entry-in

priscilawebdev committed:
feat(grouptombstones): Add 'last_seen' and 'times_seen' columns

priscilawebdev committed:
Merge branch 'master' into priscila/feat/grouptombstone/add-new-seen-columns

priscilawebdev committed:
make columns accept null

priscilawebdev committed:
revert not needed change

JoshFerge reviewed (changes requested):
Hi! thanks for this PR. this will be super helpful for users who use the discard feature. I am worried that for a frequently occuring group tombstone, incrementing and updating time seen on every event may cause a large amount of load on the database.

could we use the buffer system similar to what we do for existing groups?

https://github.com/getsentry/sentry/blob/b22cd3a448d511588dbb581b8be3b494457eae3b/src/sentry/event_manager.py#L1824

wedamija commented on src/sentry/models/grouptombstone.py:
```diff
@@ -39,6 +39,8 @@ class GroupTombstone(Model):
         blank=True, null=True
     )
     actor_id = BoundedPositiveIntegerField(null=True)
+    times_seen = BoundedPositiveIntegerField(default=0, null=True)
+    last_seen = models.DateTimeField(default=timezone.now, null=True)
```
Do you actually want these to be null by default? You can use `db_default` if you want to give existing rows a value

priscilawebdev replied:
oh no. that is not what I want. I would like existing rows to not have any value. Hopfully the buffer_inc func deals with null values

wedamija commented on src/sentry/event_manager.py:
```diff
@@ -510,7 +511,23 @@ def save_error_events(
+            if tombstone_id:
+                try:
+                    group_tombstone = GroupTombstone.objects.get(id=tombstone_id)
+                    group_tombstone.last_seen = max(
+                        job["event"].datetime, group_tombstone.last_seen
+                    )
+                    group_tombstone.times_seen += 1
```
Could there be a high volume group here that is discarded? If so, we could end up hammering postgres for each event that hits that group. Usually we use buffers for this kind of update around groups

priscilawebdev replied:
I wasn't aware of the buffer - thanks for the feedback, it makes perfect sense. I've incorporated the feedback in the PR: https://github.com/getsentry/sentry/pull/93685

wedamija reviewed (commented):
I'd suggest that we split out the migration separately, so that reverts are easier if something goes wrong

priscilawebdev commented:
> I'd suggest that we split out the migration separately, so that reverts are easier if something goes wrong

Thanks for the great feedback. I have addressed that in the PR https://github.com/getsentry/sentry/pull/93682

priscilawebdev commented:
closing this one as I have addressed all feedback in separate PRs
